src/gen_smtp_server_session.erl

```diff
@@ -113,8 +113,16 @@ start(Socket, Module, Options) ->
 %% @private
 -spec init(Args :: list()) -> {'ok', #state{}, ?TIMEOUT} | {'stop', any()} | 'ignore'.
 init([Socket, Module, Options]) ->
-	{ok, {PeerName, _Port}} = socket:peername(Socket),
-	case Module:init(proplists:get_value(hostname, Options, smtp_util:guess_FQDN()), proplists:get_value(sessioncount, Options, 0), PeerName, proplists:get_value(callbackoptions, Options, [])) of
+	PeerName = case socket:peername(Socket) of
+		{ok, {IPaddr, _Port}} -> IPaddr;
+		{'error', _} -> 'error'
+	end,
+	case PeerName =/= 'error'
+		andalso Module:init(proplists:get_value(hostname, Options, smtp_util:guess_FQDN()), proplists:get_value(sessioncount, Options, 0), PeerName, proplists:get_value(callbackoptions, Options, []))
+	of
+		false ->
+			socket:close(Socket),
+			ignore;
 		{ok, Banner, CallbackState} ->
```

Comment:
The 'andalso' suggests a boolean check, I would prefer to have the PeerName check and the init return handled separately.  In this case you can have an `init_peername({error, _}, ...)` that closes the socket and a `init_peername({ok, ...}, ...)` that handles the init.